pkg/webhook/cluster/mutation/mutation.go

```diff
@@ -0,0 +1,116 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutation
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	ctrlruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// AdmissionHandler for mutating Kubermatic Cluster CRD.
+type AdmissionHandler struct {
+	log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// NewAdmissionHandler returns a new cluster mutation AdmissionHandler.
+func NewAdmissionHandler() *AdmissionHandler {
+	return &AdmissionHandler{}
+}
+
+func (h *AdmissionHandler) InjectLogger(l logr.Logger) error {
+	h.log = l.WithName("cluster-mutation-handler")
+	return nil
+}
+
+func (h *AdmissionHandler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *AdmissionHandler) Handle(ctx context.Context, req webhook.AdmissionRequest) webhook.AdmissionResponse {
+	cluster := &kubermaticv1.Cluster{}
+	oldCluster := &kubermaticv1.Cluster{}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		// NOP
+	case admissionv1.Update:
+		if err := h.decoder.Decode(req, cluster); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if err := h.decoder.DecodeRaw(req.OldObject, oldCluster); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if err := h.mutateUpdate(ctx, oldCluster, cluster); err != nil {
+			h.log.Info("cluster mutation failed", "error", err)
+			return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("cluster mutation request %s failed: %v", req.UID, err))
+		}
+
+		mutatedCluster, err := json.Marshal(cluster)
+		if err != nil {
+			return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("marshaling cluster object failed: %v", err))
+		}
+
+		return admission.PatchResponseFromRaw(req.Object.Raw, mutatedCluster)
+	case admissionv1.Delete:
+		// NOP
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("%s not supported on cluster resources", req.Operation))
+	}
+	return webhook.Allowed(fmt.Sprintf("no mutation done for request %s", req.UID))
+}
+
+func (h *AdmissionHandler) mutateUpdate(ctx context.Context, oldCluster, newCluster *kubermaticv1.Cluster) error {
+	// This part of the code handles the CCM/CSI migration. It currently works
+	// only for OpenStack clusters, in the following way:
+	//   * Add the CCM/CSI migration annotations
+	//   * Enable the UseOctaiva flag
+	switch {
+	case newCluster.Spec.Cloud.Openstack != nil:
+		if v, ok := newCluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider]; ok && v {
+			if oldV, oldOk := oldCluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider]; !oldOk || !oldV {
```

Comment:
As the values in the feature map are bool and zero value is `false` you can avoid checking for the presence of the key:

```suggestion
		if v, oldV := newCluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider], oldCluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider]; v && !oldV {
```